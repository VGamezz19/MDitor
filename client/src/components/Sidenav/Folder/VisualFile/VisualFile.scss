@import '../../../../sass/_variables';
@import '../../../../sass/_extends';
.files {
    @extend %rectangle-container;
    border-bottom: unset;
    justify-content: space-between;
    padding: 0 2em;
    i {
        color: $main-font-color;
    }
}

.li-item-file {
    @extend %flex-in-column;
    position: relative;
    align-items: flex-end;
    a {
        padding: 0 !important;
    }
    .button-delete {
        position: absolute !important;
        top: 10px;
    }
    .button-update {
        position: absolute !important;
        right: 45px;
        top: 13px;
    }
    .button-delete:hover {
        div {
            svg {
                color: $delete-color !important;
            }
        }
    }
    .button-update:hover {
        div {
            svg {
                color: $update-color !important;
            }
        }
    }
    .files {
        @extend %rectangle-container;
        border-bottom: 1px solid $main-border-color !important;
        border-top: unset !important;
    }
}

.focus-file {
    border: 3px solid #1367ea;
}