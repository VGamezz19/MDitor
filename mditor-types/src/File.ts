/**
 *
 * DataType File, composite pattern
 * https://en.wikipedia.org/wiki/Composite_pattern
 *
 * @version 1.0.0
 */
class File {
    private id: string;
    private type: "folder" | "file";
    public title: string;
    public files?: Array<File>;
    public content?: string;

    constructor(_type: "folder" | "file", _id: string, _title: string, _content?: string, _files?: Array<File>) {

        if (_type !== "folder" && _type !== "file") {

            this.throwError("new File(type: string, id: string, title: string, content?: string, files?: File)", "only accepts 'folder' || 'file'");
        }

        if (_type == "folder") {

            this.files = _files || [];
        }

        if (_type == "file") {

            this.content = _content || "";
        }

        this.id = _id;
        this.title = _title;
        this.type = _type;
    }

    /**
     *
     * method getId
     *
     * Compatibility with File and Folder Type
     *
     * @returns {String} get this.id
     *
     * @version 1.0.0
     */
    getId(): string { return this.id; }

    /**
     *
     * method getTitle
     *
     * Compatibility with File and Folder Type
     *
     * @returns {String} get this.title
     *
     * @version 1.0.0
     */
    getTitle(): string { return this.title; }

    /**
     *
     * method getType
     *
     * Compatibility with File and Folder Type
     *
     * @returns {String} get this.type
     *
     * @version 1.0.0
     */
    getType(): string { return this.type; }

    /**
     *
     * method getContent
     *
     * Compatibility only with File Type
     *
     * @returns {String} get this.content
     *
     * @throws {Error} - mandatory Type File
     *
     * @version 1.0.0
     */
    getContent(): string | never { return this.type === "file" ? this.content : this.throwError("is not a function", this.getContent.name); }

    /**
     *
     * method getFile
     *
     * Compatibility only with Folder Type
     *
     * @param {String} id - id from retrieved file
     *
     * @returns {File} get retrieve file
     *
     * @throws {Error} - mandatory Type Folder
     *
     * @version 1.0.0
     */
    getFile(id: string): File | never { return this.type === "folder" ? this.files.find(file => file.id === id) : this.throwError("is not a function", this.getFile.name); }

    /**
     *
     * method setTitle
     *
     * Compatibility with Folder and File Type
     *
     * @param {String} title - set title
     *
     * @version 1.0.0
     */
    setTitle(title: string) { this.title = title; }

    /**
     *
     * method getFile
     *
     * Compatibility only with File Type
     *
     * @param {String} id - id from retrieved file
     *
     * @returns {File} get retrieve file
     *
     * @throws {Error} - mandatory Type File
     *
     * @version 1.0.0
     */
    setContent(content: string) { this.type === "file" ? this.content = content : this.throwError("is not a function", this.setContent.name); }

    /**
     *
     * method add
     *
     * Compatibility only with Folder Type
     *
     * @param {File} childr  children file type file to add
     *
     * @throws {Error} - mandatory Type Folder
     *
     * @version 1.0.0
     */
    add(childr: File) { this.type === "folder" ? this.files.push(childr) : this.throwError("is not a function", this.add.name); }

    /**
     *
     * method remove
     *
     * Compatibility only with Folder Type
     *
     * @param {File} child  children file type file to remove
     *
     * @throws {Error} - mandatory Type Folder
     *
     * @version 1.0.0
     */
    remove(child: File) {
        if (this.type === "folder") {
            const length = this.files.length;

            for (let i = 0; i < length; i++) {
                if (this.files[i] === child) {
                    this.files.splice(i, 1);
                    return;
                }
            }
        } else {
            this.throwError("is not a function", this.remove.name);
        }

    }

    /**
     *
     * method throwError
     *
     * @param {String} message descriptive message about error
     * @param {String} type type error or name error
     */
    throwError(message: string, type: string): never { throw new Error("Error: " + type + " " + message); }
}

export default File;