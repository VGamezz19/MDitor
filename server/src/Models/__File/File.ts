// class File {
//     private title: string;
//     private content?: string;

//     constructor(_title: string, _content?: string)Â {
//         this.title=_title;
//         this.content=_content || "";
//     }
// }

// export { File }