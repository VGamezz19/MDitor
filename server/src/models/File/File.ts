import { Schema, Model, model } from "mongoose";
import { IFileModel } from "./IFileModel.interface";

/**
 * File Schema, Object Mongoose (Object)
 *
 * Defines Schema to use in collection File
 *
 * @version 1.0.0
 */
export const FileSchema: Schema = new Schema({
  title: String,
  content: String,
  folder: { type: Schema.Types.ObjectId, ref: "Folder" }
});

export const File: Model<IFileModel> = model<IFileModel>("File", FileSchema);