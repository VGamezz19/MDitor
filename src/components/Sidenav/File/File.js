import React, { Component } from 'react'
import IconButton from 'material-ui/IconButton';
import Delete from 'material-ui/svg-icons/action/delete';


class File extends Component {
    handlerClickFile = () => {
        console.log('click FILE')
    }
    handllerClickDelete = () => {
        const idFoler = this.props.idFolder
        const idFile = this.props.idFile
 
        this.props.deleteFile(idFoler, idFile)
    }
    render() {
        return (
            <li className='li-item-file'>
                <button type='submit' className="files grey lighten-3" onClick={this.handlerClickFile}>
                    {this.props.title}
                </button>
                <IconButton className='button-delete' onClick={this.handllerClickDelete}>
                    <Delete />
                </IconButton>
            </li>
        )
    }
}

export default File