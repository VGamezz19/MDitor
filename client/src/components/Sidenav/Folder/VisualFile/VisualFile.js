import React, { Component } from 'react'
import IconButton from 'material-ui/IconButton';
import Delete from 'material-ui/svg-icons/action/delete';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { Link, Route } from "react-router-dom"
import PropTypes from 'prop-types';
import File from 'mditor-types';

class VisualFile extends Component {
    render() {

        const { file, folderId, logicFile, handleToggleSidenav } = this.props
        const fileId = file.getId();
        const fileTitle = file.getTitle();
        return (
            <Route
                path={`/${folderId}/${fileId}/(view|edit)`}
                children={({ match }) => (
                    <li className={`li-item-file ${match ? 'focus-file' : false}`}>
                        <Link 
                            className='files grey lighten-3' 
                            to={`/${folderId}/${fileId}/edit`}
                            onClick={() => handleToggleSidenav()}>
                            <button type='submit' className="files grey lighten-3">
                                {fileTitle}
                            </button>
                        </Link>
                        <MuiThemeProvider>
                            <IconButton className='button-delete' onClick={() => logicFile.delete(folderId, fileId)}>
                                <Delete />
                            </IconButton>
                        </MuiThemeProvider>
                    </li>
                )} />
        )
    }
}

VisualFile.propTypes = {
    /**
     * mandatory type
     * File from ('mditor-types');
     */
    file: PropTypes.instanceOf(File),
    /**
     * Object with logic File
     */
    logicFile: PropTypes.objectOf(PropTypes.func.isRequired)
}

export default VisualFile