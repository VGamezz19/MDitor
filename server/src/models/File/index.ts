import { IFileModel } from "./IFileModel.interface";

import { File } from "./File";

export { IFileModel, File };