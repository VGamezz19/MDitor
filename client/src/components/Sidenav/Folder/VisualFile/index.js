import './VisualFile.scss'
import VisualFile from './VisualFile'

export default VisualFile