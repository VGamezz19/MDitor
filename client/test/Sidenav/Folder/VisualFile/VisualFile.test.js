import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import VisualFile from '../../../../src/components/Sidenav/Folder/VisualFile';
import File from 'mditor-types';

const file = new File("file", 3, "title file", "some content")
const folderId = 3
const logicFile = {
    create: () => console.log("Create file"),
    delete: () => console.log("Delete file")
}


it('Main aplication rendering', () => {

    const div = document.createElement('div');

    const wrapper = <MemoryRouter>
        <VisualFile
            file={file}
            folderId={folderId}
            logicFile={logicFile}
            handleToggleSidenav={() => console.log("handle toggle sidenav")} />
    </ MemoryRouter>

    ReactDOM.render(wrapper, div);

    ReactDOM.unmountComponentAtNode(div);
});

it('Should create and Match same Snapshot', () => {

    const wrapper = <MemoryRouter>
        <VisualFile
            file={file}
            folderId={folderId}
            logicFile={logicFile}
            handleToggleSidenav={() => console.log("handle toggle sidenav")} />
    </ MemoryRouter>

    const component = renderer.create(wrapper);

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});