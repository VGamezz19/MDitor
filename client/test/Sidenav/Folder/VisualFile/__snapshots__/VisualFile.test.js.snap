// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create and Match same Snapshot 1`] = `
<li
  className="li-item-file false"
  onClick={[Function]}
>
  <a
    className="files grey lighten-3\\""
    href="/3/3/edit"
    onClick={[Function]}
  >
    <button
      className="files grey lighten-3"
      type="submit"
    >
      title file
    </button>
  </a>
  <button
    className="button-delete"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOut={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "background": "none",
        "border": 10,
        "boxSizing": "border-box",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": 0,
        "fontWeight": "inherit",
        "height": 48,
        "margin": 0,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "visible",
        "padding": 12,
        "position": "relative",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "verticalAlign": null,
        "width": 48,
      }
    }
    tabIndex={0}
    type="button"
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <svg
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fill": "currentColor",
            "height": 24,
            "muiPrepared": true,
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "width": 24,
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </button>
</li>
`;
